Implement email verification in check-out #121

rafaelgou opened the pull request:
<!-- Thanks for your pull request.  Comments within these type of tags will be hidden and not show up when you submit your pull request -->
<!-- Please answer the following questions in order to expediate acceptance -->

## Problem this Pull Request solves
<!-- Please describe your changes in the context of the problem they solve -->

**Implement "Confirm Email" question to allow email verification in check-out**

This PR implements a new system question **Confirm Email** (`emailc`). It's part of **Personal Information** group. It's optional for the end-user to use, but once activated is required.

The question is not saved, it serves only for "equalsTo" validation on the frontend.

## How has this been tested

<!-- Please describe in detail how you tested your changes and how testing can be reproduced -->
<!-- Include details of your testing environment, and tests ran to see how your changes affect other areas of code -->
<!-- Include any notes about automated tests you've written for this pull request.  Pull requests with automated tests are preferred. -->

The new field is expected to be present in _new instalations_ (see the Migration below).

So far the feature can be checked on List Question, Edit Question, Edit Group (Personal Information), and during checkout on the Registration Form.

![ee-confirm-email-edit-question](https://user-images.githubusercontent.com/196133/76254796-d67f3c80-622b-11ea-81d9-e69b9fe6f8a2.png)

![ee-confirm-email-edit-personal-group-question](https://user-images.githubusercontent.com/196133/76254802-da12c380-622b-11ea-89d5-6217e6463f44.png)

![ee-confirm-email-registration-form](https://user-images.githubusercontent.com/196133/76254812-dda64a80-622b-11ea-86cf-5002826247f2.png)

## Migration

My current understanding is to create the new field  under `schema_changes_after_migration()`, setting up a dummy `_migrate_old_row()` to fulfill the abstract method. I also understand the need of the method `can_migrate_from_version()`, but it's a little obscure to define the correct version.

Is this approach correct? Any further suggestions?

## Unit tests

I was unable to run unit tests. After installing PHPUNit v7.5.20 (for Wordpress compatibility) and run `tests/bin/install-wp-tests.sh` I got:

```bash
tests$ phpunit -v
Installing...
Running as single site... To run multisite, use -c tests/phpunit/multisite.xml
Not running ajax tests. To execute these, use --group ajax.
Not running ms-files tests. To execute these, use --group ms-files.
Not running external-http tests. To execute these, use --group external-http.
PHP Fatal error:  Declaration of SpeedTrapListener::addError(PHPUnit_Framework_Test $test, Exception $e, $time) must be compatible with PHPUnit\Framework\TestListener::addError(PHPUnit\Framework\Test $test, Throwable $t, float $time): void in /home/rafael/Dev/EventEspresso/wp/web/app/plugins/event-espresso/tests/includes/speed-trap-listener.php on line 7

Fatal error: Declaration of SpeedTrapListener::addError(PHPUnit_Framework_Test $test, Exception $e, $time) must be compatible with PHPUnit\Framework\TestListener::addError(PHPUnit\Framework\Test $test, Throwable $t, float $time): void in /home/rafael/Dev/EventEspresso/wp/web/app/plugins/event-espresso/tests/includes/speed-trap-listener.php on line 7
```

I've found some information about this incompatible signatures here: https://github.com/sebastianbergmann/phpunit/issues/2765

My question is: which version of PHPUnit should I use to run the tests?

## Checklist

* [x] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [x] User input is adequately validated and sanitized
* [x] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [x] My code follows the Event Espresso code style.
* [x] My code has proper inline documentation.
* [ ] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

tn3rb commented:
Hi Rafael, 

Thanks for submitting your PR.

We're running PHPUnit 6.5.14 on Travis, and your tests pass there except for some linting issues in 13393.4:
https://travis-ci.com/eventespresso/event-espresso-core/builds/152533330

I'll admit there hasn't been too too much PHP activity lately because I've been immersed in a React based project for ages now, so some of our PHP dependencies have obviously gotten a little behind.

I'll create an issue for updating things.

Will also try to get your PR reviewed sometime within the next couple of hours.

joshfeck commented on core/libraries/form_sections/inputs/EE_Email_Confirm_Input.input.php:
```diff
@@ -1,10 +1,11 @@
  * @package         Event Espresso
  * @subpackage
- * @author              Mike Nelson
+ * @since           4.10.4
```
You can put `$VID:$` in for new code. When the release is built the version number will get automatically swapped in.

joshfeck commented on core/helpers/EEH_Activation.helper.php:
```diff
@@ -1079,7 +1095,7 @@ public static function initialize_system_questions()
                     $QST_ID = $wpdb->insert_id;
                     // QUESTION GROUP QUESTIONS
-                    if (in_array($QST_system, array('fname', 'lname', 'email'))) {
+                    if (in_array($QST_system, array('fname', 'lname', 'email', 'emailc'))) {
```
I think we were planning on making this something the site admin would opt-in to by assigning the question to the question group.

tn3rb commented on core/libraries/form_sections/strategies/validation/EE_Equal_To_Validation_Strategy.strategy.php:
```diff
@@ -0,0 +1,61 @@
+ * @package               Event Espresso
+ * @subpackage            core
+ * @author                Mike Nelson
+ * @since                 4.6
```
# AHA!!!

I had my suspicions that you were actually Michael Nelson in disguise!!!

Nah just being silly, but you should give yourself credit for this code even if it is modelled after code that Michael wrote.
As well, for any `@since` tags for new code, we use `$VID:$` for the version number which will get automagically replaced by the correct version when a release is built.

```suggestion
 * @author                Rafael Goulart
 * @since                 $VID:$
```

tn3rb commented on core/libraries/form_sections/strategies/validation/EE_Equal_To_Validation_Strategy.strategy.php:
```diff
@@ -0,0 +1,61 @@
+    public function __construct($validation_error_message = '', $compare_to)
+    {
+        if (! $validation_error_message) {
+            $validation_error_message = esc_html__('Confirm email doesn\'t match.', 'event_espresso');
```
since you've named this validation strategy in an abstract enough way that it could apply to anything (which is great), 
let's change this error message to also be generic... maybe something like:

```suggestion
            $validation_error_message = esc_html__('Fields do not match.', 'event_espresso');
```

or if you want to get fancy, you could require the "presentation" label for the `$compare_to` field as a parameter and then `sprintf() ` that into the error message.

tn3rb commented on core/helpers/EEH_Activation.helper.php:
```diff
@@ -966,6 +967,21 @@ public static function initialize_system_questions()
+                        $QST_values = array(
+                            'QST_display_text'  => __('Confirm Email Address', 'event_espresso'),
+                            'QST_admin_label'   => __('Confirm Email Address - System Question', 'event_espresso'),
+                            'QST_system'        => 'emailc',
```
```suggestion
                            'QST_system'        => 'email_confirm',
```

tn3rb commented on modules/single_page_checkout/reg_steps/attendee_information/EE_SPCO_Reg_Step_Attendee_Information.class.php:
```diff
@@ -1227,6 +1227,11 @@ private function _save_registration_form_input(
         $input_value = ''
     ) {
+        // If emailc is sent it's not saved
+        if ((string) $form_input === 'emailc') {
```
```suggestion
        if ((string) $form_input === 'email_confirm') {
```

tn3rb commented:
here's the linting issues that also need to be corrected:

```
FILE: ...registration_form/templates/questions_main_meta_box.template.php
----------------------------------------------------------------------
FOUND 5 ERRORS AFFECTING 4 LINES
----------------------------------------------------------------------
 125 | ERROR | [x] Whitespace found at end of line
     |       |     (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 126 | ERROR | [x] Expected 0 spaces after opening bracket; newline
     |       |     found
     |       |     (PSR2.ControlStructures.ControlStructureSpacing.SpacingAfterOpenBrace)
 126 | ERROR | [x] Whitespace found at end of line
     |       |     (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 127 | ERROR | [x] Whitespace found at end of line
     |       |     (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
 130 | ERROR | [x] Array keys must be surrounded by spaces unless they
     |       |     contain a string or an integer.
     |       |     (WordPress.Arrays.ArrayKeySpacingRestrictions.NoSpacesAroundArrayKeys)
----------------------------------------------------------------------
PHPCBF CAN FIX THE 5 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
FILE: .../libraries/form_sections/inputs/EE_Email_Confirm_Input.input.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 32 | ERROR | [x] Multi-line function call not indented correctly;
    |       |     expected 16 spaces but found 12
    |       |     (PSR2.Methods.FunctionCallSignature.Indent)
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
FILE: ...rategies/validation/EE_Equal_To_Validation_Strategy.strategy.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 61 | ERROR | [x] The closing brace for the class must go on the next
    |       |     line after the body
    |       |     (PSR2.Classes.ClassDeclaration.CloseBraceAfterBody)
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
FILE: ...n_scripts/4_10_0_stages/EE_DMS_4_10_0_Confirm_Email.dmsstage.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 64 | ERROR | [x] Blank line found at start of control structure
    |       |     (Squiz.WhiteSpace.ControlStructureSpacing.SpacingAfterOpen)
 85 | ERROR | [x] The closing brace for the class must go on the next
    |       |     li
```

rafaelgou commented:
@tn3rb , @joshfeck, thanks for the review. 

- I'll try to unit test using PHPUnit 6.5.14 and give feedback.
- `email_ confirm` was my first option, but I went to `emailc` to keep short. But I agree with you, @tn3rb , being explicit avoids misreading.
- I already notice that I deserve the credit for that code, haha.
- My linting wasn't working properly, so I just did "eye's linting". 
- Thank you both for the `$VID:$` tip. Lifesaver.

It's late evening here in Brazil, I'll make the changes tomorrow morning and update the PR.

Thanks!

rafaelgou commented:
@tn3rb , @joshfeck ,

I fixed the linter issues and changed the question name from `emailc` to `email_confirm`.

I also able to run the tests using PHPUnit 6.5.*

I'd appreciate your help with the migration, that's the most obscure part for me and I'm not able to test it. It's supposed to create the new question in running installations respecting the `QST_order`.

And, of course, anything else you want to comment.

Thanks!

tn3rb commented on core/libraries/form_sections/strategies/validation/EE_Equal_To_Validation_Strategy.strategy.php:
```diff
@@ -0,0 +1,60 @@
+ * @package               Event Espresso
+ * @subpackage            core
+ * @author                Mike Nelson
+ * @since                 4.6
```
still need to tweak this phpdoc header

the class name in the comment is wrong and you're still crediting Michael :)

rafaelgou replied:
Done!

tn3rb commented on core/data_migration_scripts/4_10_0_stages/EE_DMS_4_10_0_Confirm_Email.dmsstage.php:
```diff
@@ -62,10 +62,9 @@ public function schema_changes_after_migration()
             // First checks if the field already exists
             $SQL_exists = "SELECT COUNT(*) FROM " . $question_table;
-            $exists = $wpdb->get_var($$SQL_exists);
+            $exists = $wpdb->get_var($SQL_exists);
```
OOF   totally missed that in my first review :facepalm:

tn3rb commented:
re:
> I'd appreciate your help with the migration, that's the most obscure part for me and I'm not able to test it.

the code you added _looks_ okay to me, as in I don't see anything that looks wrong when just looking at it here... 
was there something specific that was giving you problems?

rafaelgou commented:
@tn3rb I need so help to figure it out _how_ to test it. My thought is:

- Download (or in this case, clone) the last version from the repo.
- Install in an empty wp instance.
- Add some dummy data.
- Override with my pull-request data.

Right here, what should I expect to see? As far as I understand, the code is supposed to discover the database migration need and run it. Or will the user be prompted (and put in maintenance mode)?

Is that a good path to test migrations?

tn3rb commented on core/data_migration_scripts/4_10_0_stages/EE_DMS_4_10_0_Confirm_Email.dmsstage.php:
```diff
@@ -0,0 +1,87 @@
+ * @author              Brent Christensen
+ *
+ */
+class EE_DMS_4_10_0_Confirm_Email extends EE_Data_Migration_Script_Stage_Table
```
this entire file can actually be deleted...
we don't need it.

A Data Migration Script (DMS) is needed when we need to either:

- move existing data to a new table and/or field 

- modify existing data in some manner

In this PR, the new email confirmation question gets added as part of the database initialization process that you added in `EEH_Activation::initialize_system_questions()`.
I've checked out this branch and tested locally and the question data gets successfully added to the database without this DMS even running...

...you're probably wondering right now what I mean by "without this DMS even running" right?

In order for a DMS script to actually run during an update a couple of things have to happen:

1. the version number for the application has to be incremented, ex: 4.10.4 => 4.10.5

2. the DMS class has to be instantiated and added to the `_migration_stages` array for the `EE_Data_Migration_Script_Base` class of the current version, which in this case would be in the constructor of `core/data_migration_scripts/EE_DMS_Core_4_10_0.dms.php`.

an example of the latter would be:

```php
        $this->_migration_stages = array(
            new EE_DMS_4_10_0_Event_Question_Group(),
            new EE_DMS_4_10_0_Confirm_Email(),
        );
```

but again, we don't even need that because the new question is already added to the db upon activation.

joshfeck commented:
This should be tested on a new install & on an existing install. 

Also need to test the functionality:

- [ ] Ability to edit the question
- [ ] re-order the questions in the question group where the new question is added
- [ ] test main functionality (matching email fields verification)
- [ ] Can form submit after correcting a validation failure?
- [ ] test with reCAPTCHA activated

sethshoultes commented:
I can't seem to edit and save the question but everything else seems to be working.

joshfeck commented:
@sethshoultes that last commit will fix not being able to edit/save the question.

joshfeck commented:
The form validation that checks for an empty field displays what's set for `QST_required_text`, would it make sense for the same message to be used if matching validation fails?

sethshoultes commented:
This is working now. I can edit the question now.

sethshoultes commented:
> The form validation that checks for an empty field displays what's set for `QST_required_text`, would it make sense for the same message to be used if matching validation fails?

Hmm. I think it makes sense to display the current message (if that's what you are discussing):
![image](https://user-images.githubusercontent.com/235315/76991871-ef64ad80-690f-11ea-9c2b-358fa9b7cde6.png)
